service/tracker/execution_internal_test.go

```diff
@@ -0,0 +1,310 @@
+
+		_, err := NewExecution(log, db, stream)
+
+		assert.Error(t, err)
+	})
+
+	t.Run("handles missing header", func(t *testing.T) {
+		log := zerolog.Nop()
+		stream := mocks.BaselineRecordStreamer(t)
+
+		db := helpers.InMemoryDB(t)
+		require.NoError(t, db.Update(operation.InsertRootHeight(header.Height)))
+		require.NoError(t, db.Update(operation.IndexBlockHeight(header.Height, blockID)))
+		// Do not insert header.
+		//require.NoError(t, db.Update(operation.InsertHeader(blockID, header)))
+		require.NoError(t, db.Update(operation.IndexBlockSeal(blockID, seal.ID())))
+		require.NoError(t, db.Update(operation.InsertSeal(seal.ID(), seal)))
+
+		_, err := NewExecution(log, db, stream)
+
+		assert.Error(t, err)
+	})
+
+	t.Run("handles missing seal index", func(t *testing.T) {
+		log := zerolog.Nop()
+		stream := mocks.BaselineRecordStreamer(t)
+
+		db := helpers.InMemoryDB(t)
+		require.NoError(t, db.Update(operation.InsertRootHeight(header.Height)))
+		require.NoError(t, db.Update(operation.IndexBlockHeight(header.Height, blockID)))
+		require.NoError(t, db.Update(operation.InsertHeader(blockID, header)))
+		// Do not insert seal ID.
+		//require.NoError(t, db.Update(operation.IndexBlockSeal(blockID, seal.ID())))
+		require.NoError(t, db.Update(operation.InsertSeal(seal.ID(), seal)))
+
+		_, err := NewExecution(log, db, stream)
+
+		assert.Error(t, err)
+	})
+
+	t.Run("handles missing seal", func(t *testing.T) {
+		log := zerolog.Nop()
+		stream := mocks.BaselineRecordStreamer(t)
+
+		db := helpers.InMemoryDB(t)
+		require.NoError(t, db.Update(operation.InsertRootHeight(header.Height)))
+		require.NoError(t, db.Update(operation.IndexBlockHeight(header.Height, blockID)))
+		require.NoError(t, db.Update(operation.InsertHeader(blockID, header)))
+		require.NoError(t, db.Update(operation.IndexBlockSeal(blockID, seal.ID())))
+		// Do not insert seal.
+		//require.NoError(t, db.Update(operation.InsertSeal(seal.ID(), seal)))
+
+		_, err := NewExecution(log, db, stream)
+
+		assert.Error(t, err)
+	})
+}
+
+func TestExecution_Purge(t *testing.T) {
+	blockIDs := mocks.GenericBlockIDs(4)
+	var blocks []*uploader.BlockData
+	for idx := range blockIDs {
+		blocks = append(blocks, &uploader.BlockData{
+			Block: &flow.Block{
+				Header: &flow.Header{
+					Height: uint64(idx), // Block height is equal to its index in the slice.
+				},
+			},
+		})
+	}
+
+	tests := []struct {
+		name string
+
+		threshold uint64
+		before    map[flow.Identifier]*uploader.BlockData
+
+		after map[flow.Identifier]*uploader.BlockData
+	}{
+		{
+			name: "nominal case",
+
+			threshold: 2,
+			before: map[flow.Identifier]*uploader.BlockData{
+				blockIDs[0]: blocks[0],
+				blockIDs[1]: blocks[1],
+				blockIDs[2]: blocks[2],
+				blockIDs[3]: blocks[3],
+			},
+
+			after: map[flow.Identifier]*uploader.BlockData{
+				blockIDs[2]: blocks[2],
+				blockIDs[3]: blocks[3],
+			},
+		},
+		{
+			name: "purge threshold vastly over max height",
```

Comment:
Then maybe not vastly over, but at least having a test that checks what happens if we try to purge a height above the max height that we currently have makes sense IMO, on top of the three cases that you mentioned.

Also, following that logic, it would make sense to have one for the case where it's one below the lowest height we have, to make sure that we don't have a weird behavior in that case as well.

This way we have:

* middle threshold
* lowest threshold
* lowest + 1
* lowest - 1
* highest threshold
* highest + 1
* highest - 1

WDYT?

I'm also going to add a test I wanted initially but forgot about, which is to ensure that given an already empty map to purge, nothing happens.

Reply:
That is significantly more effort than required. You only need the _boundary conditions_. I recommend that you read up on this for software testing again. It means that we should have the extremes of each domain, and every boundary value.

The domain here is pretty clear, it's `[threshold:max]` for not purging, and `[min:threshold-1]` for purging. This means, simply speaking, we only need to test two values:

- threshold
- threshold-1

These two values will fully cover all domains. Now, we also want to test purging works correctly when we have multiple blocks. This means we get:

- all blocks threshold or more
- all blocks threshold-1 or less
- some blocks threshold or more, some blocks threshold-1 or less

That's all the test cases needed.